router/throttling_per_event_type_test.go

```diff
@@ -0,0 +1,374 @@
+	t.Run("throttling per event type disabled", testScenario(false, func() {
+		config.Set("Router.isolationMode", "destination")
+		config.Set("Router.throttler.adaptive.enabled", false)
+		config.Set("Router.throttler.WEBHOOK.throttlerPerEventType", false)
+	}))
+
+	// when throttling per event type is enabled, then we expect track events to be processed before identify events, because iteration will not be stopping
+	// and the rate limit for track events is higher than that for identify events
+	t.Run("static throttling with destination level isolation", testScenario(true, func() {
+		config.Set("Router.isolationMode", "destination")
+		config.Set("Router.throttler.adaptive.enabled", false)
+	}))
+	t.Run("adaptive throttling with destination level isolation", testScenario(true, func() {
+		config.Set("Router.isolationMode", "destination")
+		config.Set("Router.throttler.adaptive.enabled", true)
+	}))
+	t.Run("static throttling with workspace level isolation", testScenario(true, func() {
+		config.Set("Router.isolationMode", "workspace")
+		config.Set("Router.throttler.adaptive.enabled", false)
+	}))
+}
+
+// Using a struct to keep main_test package clean and
+// avoid function collisions with other tests
+// TODO: Move server's Run() out of main package
+type throttlingPerEventTypeMethods struct{}
+
+// newTestSpec creates a new throttlingPerEventTypeSpec with the given number of jobs per event type.
+// It generates jobs for two event types: "identify" and "track", first for "identify" and then for "track".
+func (m throttlingPerEventTypeMethods) newTestSpec(jobsPerEventType int) *throttlingPerEventTypeSpec {
+	var s throttlingPerEventTypeSpec
+	s.jobsOrdered = make([]*throttlingPerEventTypeJobSpec, jobsPerEventType*2)
+	s.receivedPerEventType = map[string]int{}
+	s.lastReceivedAt = map[string]time.Time{}
+
+	var idx int
+	for _, eventType := range []string{"identify", "track"} {
+		for i := 0; i < jobsPerEventType; i++ {
+			js := throttlingPerEventTypeJobSpec{
+				userID:    uuid.New().String(),
+				eventType: eventType,
+			}
+			s.jobsOrdered[idx] = &js
+			idx++
+		}
+	}
+	return &s
+}
+
+// newWebhook creates a new webhook server that will receive events and track their counts.
+// It will also check that the "type" field in the request body is present and is a string.
+// The server will respond with a 200 status code and the response body will contain the status code.
+// It will also update the spec with the received counts and last received time for each event type
+func (throttlingPerEventTypeMethods) newWebhook(t *testing.T, spec *throttlingPerEventTypeSpec) *throttlingPerEventTypeWebhook {
+	var wh throttlingPerEventTypeWebhook
+	wh.spec = spec
+
+	wh.Server = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		body, err := io.ReadAll(r.Body)
+		require.NoError(t, err, "should be able to read the request body")
+		typeResult := gjson.GetBytes(body, "type")
```

Comment:
We can start using the [kit jsonparser](https://github.com/rudderlabs/rudder-go-kit/blob/main/jsonparser/jsonparser.go). Once it's used everywhere we can even have rules to enforce it like we did for jsonrs.